### ISSUE TITLE
python3Packages.jaxlib: NVIDIA GPU not detected correctly

### ISSUE THREAD

**Issue by samuela** [NixOS/nixpkgs#156767]
### Describe the bug
The jax/jaxlib python packages are not detecting NVIDIA GPUs on EC2 instances running NixOS. This is due to other packages including `jaxlib` in their `propagatedBuildInputs` instead of `checkInputs`. I'll follow up with a PR to fix all of these instances.

### Steps To Reproduce
1. Run a g4dn.xlarge EC2 instance including the following in its `/etc/nixos/configuration.nix`
```nix
  # NVIDIA drivers, etc.
  nixpkgs.config.allowUnfree = true;
  services.xserver.videoDrivers = [ "nvidia" ];
  hardware.opengl.enable = true;
```
2. Set up a `shell.nix` file as follows
```nix
let
  # Last updated: 1/17/2022. Check for new commits at status.nixos.org.
  pkgs = import (fetchTarball ("https://github.com/NixOS/nixpkgs/archive/bc59ba15b64d0a0ee1d1764f18b4f3480d2c3e5a.tar.gz")) { };
in
pkgs.mkShell {
  buildInputs = with pkgs; [
    python3
    python3Packages.flax
    python3Packages.jax
    (python3Packages.jaxlib.override { cudaSupport = true; })
  ];
}
```
3. Note
```
❯ nix-shell shell.nix --run "python -c 'import jax; print(jax.devices())'"
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
[<jaxlib.xla_extension.Device object at 0x7f6a26b50270>]
```

OTOH, removing `flax` from shell.nix makes everything work just fine. This appears to be due to packages including `jaxlib` in their `propagatedBuildInputs` which creates multiple jaxlib editions in `$PYTHONPATH`:
```
❯ nix-shell shell.nix --run "echo \$PYTHONPATH | tr ':' '\n' | grep jaxlib"
/nix/store/2b9adfxqq0ix45q0qslhwwrl86z5jw97-python3.9-jaxlib-0.1.75/lib/python3.9/site-packages
/nix/store/qrddbps5cq62n2d0fg05dm73ykxzsyv3-python3.9-jaxlib-0.1.75/lib/python3.9/site-packages
```
This means the CPU version of jaxlib is clobbering the GPU-version. We need to remove `jaxlib` from the `propagatedBuildInputs` from all packages so that users have the choice of which version of jaxlib to use.

### Expected behavior
GPU to be automatically detected when the correct drivers are installed/configured.

### Screenshots
n/a

### Additional context
I initially suspected that this was related to the transition to a source-build of jaxlib in https://github.com/NixOS/nixpkgs/pull/151909, but that appears to be unrelated.

### Notify maintainers

@samuela @ndl 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.91, NixOS, 21.11 (Porcupine)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.16`
 - channels(root): `"nixos-21.11.335376.610d4ea2750, nix-ld"`
 - channels(skainswo): `"home-manager, nixpkgs-unstable-22.05pre343295.adf7f03d3bf"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by samuela** [comment]
@ndl It appears I missed this in reviewing your jax package PRs. I'll submit a PR soon!

**Comment by ndl** [comment]
Thanks for the investigation / debugging! Note that just moving `jaxlib` from `propagatedBuildInputs` to `checkInputs` might be not sufficient - IIRC I've already removed `jaxlib` in all packages that didn't explicitly require it in `setup.py`, but some packages do. I'm not sure if putting it just to `buildInputs` would work to silence the setup and avoid propagation, but if not - several packages might need patches that remove `jaxlib` from the list of their explicit requirements in `setup.py`.

In general, I'm not 100% what's the best approach here, though, as the situation with `tensorflow` is actually similar - and some packages depend on it as well. In my local setup I typically just override both `jaxlib` and `tensorflow` globally in the overlay with the GPU-enabled version, thus injecting them everywhere and consistently.

**Comment by samuela** [comment]
Oof, that's tricky. I'd like to avoid relying on overlays if at all possible, but I do see what you mean here... I'll try to see how far I can get patching the packages to avoid the explicit jaxlib dependency and go from there. IIUC they aren't supposed to be directly depending on jaxlib to begin with.

**Comment by samuela** [comment]
Does tensorflow/pytorch default `cudaSupport = true` or do they also default to not including it? These days everyone is training on GPUs, so maybe another solution here is to just default to `cudaSupport = true`.

The overlay solution would work but it requires massive rebuilds of the entire ecosystem to be done by every user.

**Comment by ndl** [comment]
For `tensorflow` see https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/tensorflow/default.nix#L16 and https://github.com/ndl/nixpkgs/blob/master/pkgs/top-level/python-packages.nix#L9629, for `pytorch` it's https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/pytorch/default.nix#L2 and https://github.com/ndl/nixpkgs/blob/master/pkgs/top-level/python-packages.nix#L8130 respectively.

TL;DR: both of them define aliases `<package>WithCuda`, `<package>WithoutCuda` and just `<package>` where `<package>` uses `pkgs.config.cudaSupport` - which, I think, is `false` by default => if somebody depends just on `<package>`, they use non-CUDA version.

Note also that the comment from TF derivation highlights the fact that enabling CUDA by default would make the packages unfree - not sure how problematic that is (= will break them for everybody who doesn't have `unfree` enabled in local config?), just wanted to mention it explicitly.
